pkg/envoy/ads/stream.go

```diff
@@ -185,41 +223,57 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
 				xdsUpdatePaths = mapset.NewSetWith(typeURL)
 			}
 
-			err = s.sendResponse(xdsUpdatePaths, proxy, &server, &discoveryRequest, s.cfg)
-			if err != nil {
-				log.Error().Err(err).Msgf("Failed to create and send %s update to Envoy with xDS Certificate SerialNumber=%s on Pod with UID=%s",
-					envoy.XDSShortURINames[typeURL], proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-				continue
-			}
+			jobDone := make(chan struct{}, 1)
+			s.workers.AddJob(&sendJob{
+				typeurls:       xdsUpdatePaths,
+				proxy:          proxy,
+				cfg:            s.cfg,
+				adsStream:      &server,
+				request:        &discoveryRequest,
+				xdsServer:      s,
+				jobDoneChannel: jobDone,
+			})
+			<-jobDone
 
 		case <-broadcastUpdate:
 			log.Info().Msgf("Broadcast wake for Proxy SerialNumber=%s UID=%s", proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-			err := s.sendResponse(mapset.NewSetWith(
-				envoy.TypeCDS,
-				envoy.TypeEDS,
-				envoy.TypeLDS,
-				envoy.TypeRDS,
-				envoy.TypeSDS),
-				proxy, &server, nil, s.cfg)
-			if err != nil {
-				log.Error().Err(err).Msgf("Failed to create and send ADS update to Envoy with xDS Certificate SerialNumber=%s on Pod with UID=%s",
-					proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-				continue
-			}
+			jobDone := make(chan struct{}, 1)
```

Comment:
Why buffer of 1?

Reply:
Worker does not block, or at least I did not think of it as blocking on done. If you don't make a buffered channel, the worker will wait for the channel to drain.
The issuer can wait for the job done if it wants to, but I did not intend to force a blocking wait for job to complete.